### PR TITLE
Avoid recursion in graph traverse

### PR DESCRIPTION
**Pull request by wschin**
It's easy to reach recursion limit in Python when calling `dfs_find_cycle` in big graphs (e.g., searching for attention heads in GPT-2 via SubgraphMatcher). Let's switch to queue-based graph tarversing.

### CODE REVIEW

File: torch/fx/passes/utils/fuser_utils.py [hunk 1]
```diff
@@ -50,19 +50,27 @@ def validate_partition(partition: NodeList) -> bool:
                 # external user node, need to expose as an output
                 outputs.append(user_node)
 
-    # perform DFS on the parition outputs
-    # if it reaches a node within the partition, then it found a cycle
+    # Set used to execlude nodes that have already been visited. 
+    # If a node has been visited, that node and all its children have
+    # been checked for cycles.
     visited: NodeSet = set()
```
Comment (Skylion007) [inline torch/fx/passes/utils/fuser_utils.py:56]: visited can be a local variable now, right? No reason for it to exist outside of the function, now?

Comment (wschin) [reply torch/fx/passes/utils/fuser_utils.py:56]: Well, it is reused for different outputs. If output `A` already check node `N`, we don't want to check `N` when processing output `B`.

Comment (wschin) [reply torch/fx/passes/utils/fuser_utils.py]: New commit pushed for improving clarify -- https://github.com/pytorch/pytorch/pull/95723/commits/2490f5f1bcd74598094322173300a395fd06cef1

File: torch/fx/passes/utils/fuser_utils.py [hunk 2]
```diff
@@ -50,23 +50,36 @@ def validate_partition(partition: NodeList) -> bool:
...
+            current = queue.pop()
+            visited.add(current)
+            if current in partition_set:
+                # Started from partition's `output` nodes, and reached
+                # another node in partition. Cycle!
+                return True
+            for user_node in current.users:
+                if user_node in visited:
+                    continue
+                queue.append(user_node)
```
Comment (SherlockNoMad) [inline torch/fx/passes/utils/fuser_utils.py:70]: lol, this is actually bfs. 
let's rename the function to bfs then.

File: torch/fx/passes/utils/fuser_utils.py [hunk 3]
```diff
@@ -50,23 +50,36 @@ def validate_partition(partition: NodeList) -> bool:
...
-            if user_node not in visited:
-                if dfs_find_cycle(user_node):
-                    return True
+    # Perform DFS on the partition outputs.
+    # If it reaches a node within the partition, then it found a cycle.
+    def dfs_find_cycle(node, visited):
+        # Start with `node` and traverse through (toward child nodes)
+        # its connected sub-graph. Nodes in `visited` won't be added
+        # to `queue` again.
+        queue: NodeList = [node]
```
Comment (SherlockNoMad) [inline torch/fx/passes/utils/fuser_utils.py:59]: nit, you can pass all the output nodes into this function in one shot, and initialize the queue with the list of nodes.

**Review by SherlockNoMad — APPROVED** [review]
per comment, and CI passing.

**Review by Skylion007 — APPROVED** [review]

**Pull request closed without merging**